pkg/compact/compact.go

```diff
@@ -234,7 +244,24 @@ func (s *Syncer) GarbageCollect(ctx context.Context) error {
 
 	begin := time.Now()
 
-	garbageIds := s.duplicateBlocksFilter.DuplicateIDs()
+	duplicateIds := s.duplicateBlocksFilter.DuplicateIDs()
+	deletionMarkIDs := s.ignoreDeletionMarkFilter.DeletionMarkBlocks(0)
+
+	// GarbageIds contains the duplicateIds, since these blocks can be replaced with other blocks.
+	// We also remove ids present in deletionMarkIDs since these blocks are already marked for deletion.
+	garbageIds := []ulid.ULID{}
+	for _, id := range duplicateIds {
+		exists := false
+		for _, deletionMarkID := range deletionMarkIDs {
+			if id.Compare(deletionMarkID) == 0 {
+				exists = true
```

Comment:
I think we can break here and avoid looping, right?

Reply:
Also let's just append straight away and then break, so we don't need `exist` bool even :hugs:

Reply:
yes, 100% agreed. Did both of these.

Reply:
I will edit the above statement about 100% agree. If we remove the exists check then the tests fail.
This is because we have to check if the duplicate id is present in deletionMarkIDs, but if the deletionMarkIDs is empty then none of the duplicate ids are considered. So we need to have the else check.
Please suggest corrections if you find any other way that is better.

Reply:
I would need to go through whole flow again, let's do this during meeting

Reply:
I might have been wrong (:

Reply:
I am returning a map now as discussed over call. Marking this thread as resolved